contrib/kafka/kafkaeventsource-operator/pkg/controller/kafkaeventsource/kafkaeventsource_controller.go

```diff
@@ -0,0 +1,327 @@
+	"k8s.io/client-go/rest"
+
+	sourcesv1alpha1 "github.com/knative/eventing-sources/contrib/kafka/kafkaeventsource-operator/pkg/apis/sources/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Add creates a new KafkaEventSource Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileKafkaEventSource{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("kafkaeventsource-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KafkaEventSource
+	err = c.Watch(&source.Kind{Type: &sourcesv1alpha1.KafkaEventSource{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secondary resource Deployments and requeue the owner KafkaEventSource
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &sourcesv1alpha1.KafkaEventSource{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileKafkaEventSource{}
+
+// ReconcileKafkaEventSource reconciles a KafkaEventSource object
+type ReconcileKafkaEventSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client        client.Client
+	dynamicClient dynamic.Interface
+	scheme        *runtime.Scheme
+}
+
+// InjectConfig implemnts an interface which means the K8s config gets injected
+func (r *ReconcileKafkaEventSource) InjectConfig(c *rest.Config) error {
+	var err error
+	r.dynamicClient, err = dynamic.NewForConfig(c)
+	return err
+}
+
+// Reconcile contains the main reconcile loop - read the state of the cluster and make changes based on the KafkaEventSource.Spec
+func (r *ReconcileKafkaEventSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
This follows a different pattern than most of the existing sources. The existing sources update an in-memory representation of the source, then after finishing (or hitting an error), it is updated once. 

It is not necessary to change to that pattern, but I think it makes it easier to see what is happening.